system/Database/BaseBuilder.php

```diff
@@ -1881,16 +1837,172 @@ public function setInsertBatch($key, string $value = '', ?bool $escape = null)
+     * @param bool            $addToDefault adds update fields to the default ones
+     * @param array|null      $ignore       ignores items in set
+     *
+     * @return $this
+     */
+    public function updateFields($set, bool $addToDefault = false, ?array $ignore = null)
+    {
+        if (! empty($set)) {
+            if (! is_array($set)) {
+                $set = explode(',', $set);
+            }
+
+            foreach ($set as $key => $value) {
+                if (! ($value instanceof RawSql)) {
+                    $value = $this->db->protectIdentifiers($value);
+                }
+
+                if (is_numeric($key)) {
+                    $key = $value;
+                }
+
+                if ($ignore === null || ! in_array($key, $ignore, true)) {
+                    if ($addToDefault) {
+                        $this->QBOptions['updateFieldsAdditional'][$this->db->protectIdentifiers($key)] = $value;
+                    } else {
+                        $this->QBOptions['updateFields'][$this->db->protectIdentifiers($key)] = $value;
+                    }
+                }
+            }
+
+            if ($addToDefault === false && isset($this->QBOptions['updateFieldsAdditional'], $this->QBOptions['updateFields'])) {
+                $this->QBOptions['updateFields'] = array_merge($this->QBOptions['updateFields'], $this->QBOptions['updateFieldsAdditional']);
+
+                unset($this->QBOptions['updateFieldsAdditional']);
+            }
+        }
+
+        return $this;
+    }
+
+    /**
+     * Sets constraints for batch upsert, update
+     *
+     * @param array|RawSql|string $set a string of columns, key value pairs, or RawSql
+     *
+     * @return $this
+     */
+    public function onConstraint($set)
+    {
+        if (! empty($set)) {
+            if (is_string($set)) {
+                $set = explode(',', $set);
+
+                $set = array_map(static fn ($key) => trim($key), $set);
+            }
+
+            if ($set instanceof RawSql) {
+                $set = [$set];
+            }
+
+            foreach ($set as $key => $value) {
+                if (! ($value instanceof RawSql)) {
+                    $value = $this->db->protectIdentifiers($value);
+                }
+
+                $this->QBOptions['constraints'][$key] = $value;
+            }
         }
 
         return $this;
     }
 
+    /**
+     * Converts value array of array to array of strings
+     */
+    protected function formatValues(array $values): array
+    {
+        return array_map(static fn ($index) => '(' . implode(',', $index) . ')', $values);
+    }
+
+    /**
+     * Compiles batch insert strings and runs the queries
+     *
+     * @param array|string|null $set a dataset or select query
+     *
+     * @return false|int|string[] Number of rows inserted or FALSE on failure, SQL array when testMode
+     */
+    public function insertBatch($set = null, ?bool $escape = null, int $batchSize = 100)
```

Comment:
`public function insertBatch(?array $set = null, ?bool $escape = null, int $batchSize = 100)`

The type for `$set` has changed. This is a BC.
Should be added to the changelog.

Reply:
I added this.. it now accepts an object of a single row of data. I also had here that it accepts a string for SQL but I haven't added all that functionality yet so I removed string from it.